File: src-tauri/src/sceneapp/catalog.rs
use super::dto::*;

const SEEDED_SCENEAPP_CATALOG_VERSION: &str = "2026-04-15";
const SEEDED_SCENEAPP_GENERATED_AT: &str = "2026-04-15T00:00:00.000Z";
const SEEDED_SCENEAPP_PACKAGE_ID: &str = "lime-core-sceneapps";

fn service_skill_binding(
    binding_family: SceneAppBindingFamily,
    service_skill_id: &str,
    skill_key: &str,
    aliases: &[&str],
) -> SceneAppEntryBinding {
    SceneAppEntryBinding {
        kind: SceneAppEntryBindingKind::ServiceSkill,
        binding_family,
        service_skill_id: Some(service_skill_id.to_string()),
        skill_key: Some(skill_key.to_string()),
        scene_key: None,
        command_prefix: None,
        aliases: aliases.iter().map(|alias| alias.to_string()).collect(),
    }
}

fn scene_binding(
    binding_family: SceneAppBindingFamily,
    scene_key: &str,
    command_prefix: &str,
    aliases: &[&str],
) -> SceneAppEntryBinding {
    SceneAppEntryBinding {
        kind: SceneAppEntryBindingKind::Scene,
        binding_family,
        service_skill_id: None,
        skill_key: None,
        scene_key: Some(scene_key.to_string()),
        command_prefix: Some(command_prefix.to_string()),
        aliases: aliases.iter().map(|alias| alias.to_string()).collect(),
    }
}

fn requirement(kind: SceneAppLaunchRequirementKind, message: &str) -> SceneAppLaunchRequirement {
    SceneAppLaunchRequirement {
        kind,
        message: message.to_string(),
    }
}

fn delivery_profile(
    artifact_profile_ref: &str,
    viewer_kind: &str,
    required_parts: &[&str],
    primary_part: &str,
) -> SceneAppDeliveryProfile {
    SceneAppDeliveryProfile {
        artifact_profile_ref: Some(artifact_profile_ref.to_string()),
        viewer_kind: Some(viewer_kind.to_string()),
        required_parts: required_parts.iter().map(|part| part.to_string()).collect(),
        primary_part: Some(primary_part.to_string()),
    }
}

fn composition_step(
    id: &str,
    order: usize,
    binding_profile_ref: &str,
    binding_family: SceneAppBindingFamily,
) -> SceneAppCompositionStepDescriptor {
    SceneAppCompositionStepDescriptor {
        id: id.to_string(),
        order,
        binding_profile_ref: Some(binding_profile_ref.to_string()),
        binding_family: Some(binding_family),
    }
}

fn composition_profile(
    blueprint_ref: &str,
    steps: Vec<SceneAppCompositionStepDescriptor>,
) -> SceneAppCompositionProfile {
    SceneAppCompositionProfile {
        blueprint_ref: Some(blueprint_ref.to_string()),
        step_count: steps.len(),
        steps,
    }
}

fn scorecard_profile(
    profile_ref: &str,
    metric_keys: &[&str],
    failure_signals: &[&str],
) -> SceneAppScorecardProfile {
    SceneAppScorecardProfile {
        profile_ref: Some(profile_ref.to_string()),
        metric_keys: metric_keys
            .iter()
            .map(|metric| metric.to_string())
            .collect(),
        failure_signals: failure_signals
            .iter()
            .map(|signal| signal.to_string())
            .collect(),
    }
}

fn seeded_story_video_suite() -> SceneAppDescriptor {
    SceneAppDescriptor {
        id: "story-video-suite".to_string(),
        title: "短视频编排".to_string(),
        summary: "把文本、线框图、配乐、剧本和短视频草稿收口成一条多模态结果链。".to_string(),
        category: "Scene Apps".to_string(),
        sceneapp_type: SceneAppType::Hybrid,
        pattern_primary: SceneAppPattern::Pipeline,
        pattern_stack: vec![
            SceneAppPattern::Pipeline,
            SceneAppPattern::Inversion,
            SceneAppPattern::Generator,
            SceneAppPattern::Reviewer,
        ],
        capability_refs: vec![
            "cloud_scene".to_string(),
            "native_skill".to_string(),
            "workspace_storage".to_string(),
            "artifact_viewer".to_string(),
        ],
        infra_profile: vec![
            "composition_blueprint".to_string(),
            "project_pack".to_string(),
            "workspace_storage".to_string(),
            "cloud_runtime".to_string(),
            "timeline".to_string(),
        ],
        delivery_contract: SceneAppDeliveryContract::ProjectPack,
        artifact_kind: Some("artifact_bundle".to_string()),
        output_hint: "短视频项目包".to_string(),
        entry_bindings: vec![
            service_skill_binding(
                SceneAppBindingFamily::CloudScene,
                "sceneapp-service-story-video",
                "story-video-suite",
                &["story-video", "mv-pipeline"],
            ),
            scene_binding(
                SceneAppBindingFamily::CloudScene,
                "story-video-suite",
                "/story-video-suite",
                &["story-video-scene"],
            ),
        ],
        launch_requirements: vec![
            requirement(
                SceneAppLaunchRequirementKind::UserInput,
                "需要主题、风格或脚本线索作为场景输入。",
            ),
            requirement(
                SceneAppLaunchRequirementKind::Project,
                "需要项目目录承接线框图、脚本和媒体结果。",
            ),
            requirement(
                SceneAppLaunchRequirementKind::CloudSession,
                "需要可用的云端运行时来完成多模态媒体处理。",
            ),
        ],
        linked_service_skill_id: Some("sceneapp-service-story-video".to_string()),
        linked_scene_key: Some("story-video-suite".to_string()),
        delivery_profile: Some(delivery_profile(
            "story-video-artifacts",
            "artifact_bundle",
            &[
                "brief",
                "storyboard",
                "script",
                "music_refs",
                "video_draft",
                "review_note",
            ],
            "brief",
        )),
        composition_profile: Some(composition_profile(
            "story-video-blueprint",
            vec![
                composition_step(
                    "brief",
                    1,
                    "story-video-native-binding",
                    SceneAppBindingFamily::NativeSkill,
                ),
                composition_step(
                    "storyboard",
                    2,
                    "story-video-native-binding",
                    SceneAppBindingFamily::NativeSkill,
                ),
                composition_step(
                    "script",
                    3,
                    "story-video-native-binding",
                    SceneAppBindingFamily::NativeSkill,
                ),
                composition_step(
                    "music_refs",
                    4,
                    "story-video-cloud-binding",
                    SceneAppBindingFamily::CloudScene,
                ),
                composition_step(
                    "video_draft",
                    5,
                    "story-video-cloud-binding",
                    SceneAppBindingFamily::CloudScene,
                ),
                composition_step(
                    "review_note",
                    6,
                    "story-video-native-binding",
                    SceneAppBindingFamily::NativeSkill,
                ),
            ],
        )),
        scorecard_profile: Some(scorecard_profile(
            "story-video-scorecard",
            &[
                "complete_pack_rate",
                "review_pass_rate",
                "publish_conversion_rate",
            ],
            &["pack_incomplete", "review_blocked", "publish_stalled"],
        )),
        aliases: vec![
            "story-video".to_string(),
            "mv-pipeline".to_string(),
            "short-video-suite".to_string(),
        ],
        source_package_id: SEEDED_SCENEAPP_PACKAGE_ID.to_string(),
        source_package_version: SEEDED_SCENEAPP_CATALOG_VERSION.to_string(),
    }
}

fn seeded_article_export() -> SceneAppDescriptor {
    SceneAppDescriptor {
        id: "x-article-export".to_string(),
        title: "网页导出".to_string(),
        summary: "在真实浏览器上下文中抓取网页正文、图片与元信息，并沉淀为项目内 Markdown 资料包。"
            .to_string(),
        category: "Scene Apps".to_string(),
        sceneapp_type: SceneAppType::BrowserGrounded,
        pattern_primary: SceneAppPattern::Pipeline,
        pattern_stack: vec![
            SceneAppPattern::Pipeline,
            SceneAppPattern::ToolWrapper,
            SceneAppPattern::Generator,
            SceneAppPattern::Inversion,
        ],
        capability_refs: vec![
            "browser_assist".to_string(),
            "workspace_storage".to_string(),
            "artifact_viewer".to_string(),
        ],
        infra_profile: vec![
            "browser_connector".to_string(),
            "site_adapter".to_string(),
            "workspace_storage".to_string(),
            "artifact_bundle".to_string(),
        ],
        delivery_contract: SceneAppDeliveryContract::ProjectPack,
        artifact_kind: Some("document".to_string()),
        output_hint: "网页资料包".to_string(),
        entry_bindings: vec![
            service_skill_binding(
                SceneAppBindingFamily::BrowserAssist,
                "sceneapp-service-article-export",
                "x-article-export",
                &["article-export"],
            ),
            scene_binding(
                SceneAppBindingFamily::BrowserAssist,
                "x-article-export",
                "/x-article-export",
                &["web-article-export"],
            ),
        ],
        launch_requirements: vec![
            requirement(
                SceneAppLaunchRequirementKind::BrowserSession,
                "需要真实网页上下文或浏览器附着会话。",
            ),
            requirement(
                SceneAppLaunchRequirementKind::Project,
                "需要项目目录来保存 Markdown 与图片资源。",
            ),
        ],
        linked_service_skill_id: Some("sceneapp-service-article-export".to_string()),
        linked_scene_key: Some("x-article-export".to_string()),
        delivery_profile: Some(delivery_profile(
            "article-export-artifacts",
            "document",
            &["index.md", "meta.json"],
            "index.md",
        )),
        composition_profile: None,
        scorecard_profile: Some(scorecard_profile(
            "article-export-scorecard",
            &["success_rate", "reuse_rate"],
            &["pack_incomplete"],
        )),
        aliases: vec![
            "article-export".to_string(),
            "web-article-export".to_string(),
        ],
        source_package_id: SEEDED_SCENEAPP_PACKAGE_ID.to_string(),
        source_package_version: SEEDED_SCENEAPP_CATALOG_VERSION.to_string(),
    }
}

fn seeded_voice_runtime() -> SceneAppDescriptor {
    SceneAppDescriptor {
        id: "voice-runtime".to_string(),
        title: "配音生成".to_string(),
        summary: "把文稿或视频内容交给托管能力完成配音，并回流成可试听的媒体产物。".to_string(),
        category: "Scene Apps".to_string(),
        sceneapp_type: SceneAppType::CloudManaged,
        pattern_primary: SceneAppPattern::Pipeline,
        pattern_stack: vec![SceneAppPattern::Pipeline, SceneAppPattern::Generator],
        capability_refs: vec![
            "cloud_scene".to_string(),
            "workspace_storage".to_string(),
            "artifact_viewer".to_string(),
        ],
        infra_profile: vec![
            "cloud_runtime".to_string(),
            "media_artifact".to_string(),
            "json_snapshot".to_string(),
        ],
        delivery_contract: SceneAppDeliveryContract::ArtifactBundle,
        artifact_kind: Some("report".to_string()),
        output_hint: "媒体结果".to_string(),
        entry_bindings: vec![service_skill_binding(
            SceneAppBindingFamily::CloudScene,
            "sceneapp-service-voice-runtime",
            "voice-runtime",
            &["voiceover", "tts-scene"],
        )],
        launch_requirements: vec![
            requirement(
                SceneAppLaunchRequirementKind::UserInput,
                "需要文稿正文或待配音素材说明。",
            ),
            requirement(
                SceneAppLaunchRequirementKind::CloudSession,
                "需要云端媒体运行时或 OEM 托管会话。",
            ),
        ],
        linked_service_skill_id: Some("sceneapp-service-voice-runtime".to_string()),
        linked_scene_key: Some("voice-runtime".to_string()),
        delivery_profile: Some(delivery_profile(
            "voice-runtime-artifacts",
            "artifact_bundle",
            &["audio.mp3", "review_note"],
            "audio.mp3",
        )),
        composition_profile: None,
        scorecard_profile: Some(scorecard_profile(
            "voice-runtime-scorecard",
            &["success_rate", "reuse_rate"],
            &["pack_incomplete"],
        )),
        aliases: vec!["voiceover".to_string(), "tts-scene".to_string()],
        source_package_id: SEEDED_SCENEAPP_PACKAGE_ID.to_string(),
        source_package_version: SEEDED_SCENEAPP_CATALOG_VERSION.to_string(),
    }
}

fn seeded_daily_trend_briefing() -> SceneAppDescriptor {
    SceneAppDescriptor {
        id: "daily-trend-briefing".to_string(),
        title: "每日趋势摘要".to_string(),
        summary: "把研究主题转成可持续运行的本地 durable 场景，并定时回流结果和失败原因。"
            .to_string(),
        category: "Scene Apps".to_string(),
        sceneapp_type: SceneAppType::LocalDurable,
        pattern_primary: SceneAppPattern::Pipeline,
        pattern_stack: vec![SceneAppPattern::Pipeline, SceneAppPattern::Reviewer],
        capability_refs: vec![
            "automation_job".to_string(),
            "workspace_storage".to_string(),
            "timeline".to_string(),
        ],
        infra_profile: vec![
            "automation_schedule".to_string(),
            "db_store".to_string(),
            "json_snapshot".to_string(),
        ],
        delivery_contract: SceneAppDeliveryContract::TableReport,
        artifact_kind: Some("table_report".to_string()),
        output_hint: "趋势摘要".to_string(),
        entry_bindings: vec![service_skill_binding(
            SceneAppBindingFamily::AutomationJob,
            "sceneapp-service-daily-trend",
            "daily-trend-briefing",
            &["trend-briefing", "growth-monitor"],
        )],
        launch_requirements: vec![
            requirement(
                SceneAppLaunchRequirementKind::Project,
                "需要工作区或项目目录保存运行历史与结果快照。",
            ),
            requirement(
                SceneAppLaunchRequirementKind::Automation,
                "需要可用的自动化调度能力。",
            ),
        ],
        linked_service_skill_id: Some("sceneapp-service-daily-trend".to_string()),
        linked_scene_key: Some("daily-trend-briefing".to_string()),
        delivery_profile: Some(delivery_profile(
            "daily-trend-artifacts",
            "table_report",
            &["brief", "review_note"],
            "brief",
        )),
        composition_profile: None,
        scorecard_profile: Some(scorecard_profile(
            "daily-trend-scorecard",
            &["success_rate", "reuse_rate"],
            &["automation_timeout"],
        )),
        aliases: vec!["trend-briefing".to_string(), "growth-monitor".to_string()],
        source_package_id: SEEDED_SCENEAPP_PACKAGE_ID.to_string(),
        source_package_version: SEEDED_SCENEAPP_CATALOG_VERSION.to_string(),
    }
}

pub fn seeded_sceneapp_descriptors() -> Vec<SceneAppDescriptor> {
    vec![
        seeded_story_video_suite(),
        seeded_article_export(),
        seeded_voice_runtime(),
        seeded_daily_trend_briefing(),
    ]
}

pub fn seeded_sceneapp_catalog() -> SceneAppCatalog {
    SceneAppCatalog {
        version: SEEDED_SCENEAPP_CATALOG_VERSION.to_string(),
        generated_at: SEEDED_SCENEAPP_GENERATED_AT.to_string(),
        items: seeded_sceneapp_descriptors(),
    }
}

pub fn get_sceneapp_descriptor(id: &str) -> Option<SceneAppDescriptor> {
    let normalized = id.trim();
    if normalized.is_empty() {
        return None;
    }

    seeded_sceneapp_descriptors()
        .into_iter()
        .find(|descriptor| {
            descriptor.id == normalized
                || descriptor
                    .aliases
                    .iter()
                    .any(|alias| alias.eq_ignore_ascii_case(normalized))
                || descriptor
                    .linked_scene_key
                    .as_deref()
                    .is_some_and(|scene_key| scene_key == normalized)
                || descriptor
                    .linked_service_skill_id
                    .as_deref()
                    .is_some_and(|service_skill_id| service_skill_id == normalized)
        })
}

File: src/lib/sceneapp/catalog.ts
import {
  readStoredBaseSetupPackageSnapshot,
  type StoredBaseSetupPackageSnapshot,
} from "@/lib/base-setup/storage";
import type {
  BaseSetupAllowedBindingFamily,
  BaseSetupArtifactProfile,
  BaseSetupBindingProfile,
  BaseSetupCatalogProjection,
  BaseSetupCompositionBlueprint,
  BaseSetupPackage,
  BaseSetupScorecardProfile,
} from "@/lib/base-setup/types";
import type {
  SceneAppCatalog,
  SceneAppCompositionProfile,
  SceneAppCompositionStepDescriptor,
  SceneAppDescriptor,
  SceneAppEntryBinding,
  SceneAppLaunchRequirement,
  SceneAppDeliveryContract,
  SceneAppDeliveryProfile,
  SceneAppPattern,
  SceneAppScorecardProfile,
  SceneAppType,
} from "./types";

interface SceneProjectionGroup {
  serviceSkillProjection?: BaseSetupCatalogProjection;
  sceneProjection?: BaseSetupCatalogProjection;
}

function normalizeCompatSceneAppBindingFamily(
  bindingFamily: BaseSetupAllowedBindingFamily,
): SceneAppEntryBinding["bindingFamily"] {
  return bindingFamily === "cloud_scene" ? "agent_turn" : bindingFamily;
}

function normalizeCompatSceneAppCapabilityRef(capabilityRef: string): string {
  return capabilityRef === "cloud_scene" ? "agent_turn" : capabilityRef;
}

function normalizeString(value?: string): string | undefined {
  const trimmed = value?.trim();
  return trimmed ? trimmed : undefined;
}

function dedupeStrings(values: Array<string | undefined>): string[] {
  return Array.from(
    new Set(
      values
        .filter((value): value is string => typeof value === "string")
        .map((value) => value.trim())
        .filter(Boolean),
    ),
  );
}

function buildProjectionGroups(
  pkg: BaseSetupPackage,
): Map<string, SceneProjectionGroup> {
  const groups = new Map<string, SceneProjectionGroup>();

  for (const projection of pkg.catalogProjections) {
    if (
      projection.targetCatalog !== "service_skill_catalog" &&
      projection.targetCatalog !== "scene_catalog"
    ) {
      continue;
    }

    const groupKey = normalizeString(projection.entryKey) ?? projection.id;
    const current = groups.get(groupKey) ?? {};

    if (projection.targetCatalog === "service_skill_catalog") {
      current.serviceSkillProjection = projection;
    } else {
      current.sceneProjection = projection;
    }

    groups.set(groupKey, current);
  }

  return groups;
}

function buildBindingProfileMap(
  pkg: BaseSetupPackage,
): Map<string, BaseSetupBindingProfile> {
  return new Map(pkg.bindingProfiles.map((profile) => [profile.id, profile]));
}

function buildArtifactProfileMap(
  pkg: BaseSetupPackage,
): Map<string, BaseSetupArtifactProfile> {
  return new Map(pkg.artifactProfiles.map((profile) => [profile.id, profile]));
}

function buildCompositionBlueprintMap(
  pkg: BaseSetupPackage,
): Map<string, BaseSetupCompositionBlueprint> {
  return new Map(
    (pkg.compositionBlueprints ?? []).map((blueprint) => [
      blueprint.id,
      blueprint,
    ]),
  );
}

function buildScorecardProfileMap(
  pkg: BaseSetupPackage,
): Map<string, BaseSetupScorecardProfile> {
  return new Map(pkg.scorecardProfiles.map((profile) => [profile.id, profile]));
}

function resolveDistinctBindingFamilies(
  blueprint: BaseSetupCompositionBlueprint | undefined,
  bindingProfileMap: Map<string, BaseSetupBindingProfile>,
): Array<SceneAppEntryBinding["bindingFamily"]> {
  if (!blueprint?.steps?.length) {
    return [];
  }

  const families = blueprint.steps
    .map((step) => step.bindingProfileRef)
    .filter((value): value is string => Boolean(value))
    .map((ref) => bindingProfileMap.get(ref)?.bindingFamily)
    .filter((value): value is BaseSetupAllowedBindingFamily => Boolean(value))
    .map((bindingFamily) =>
      normalizeCompatSceneAppBindingFamily(bindingFamily),
    );

  return Array.from(new Set(families));
}

function inferSceneAppType(
  binding: BaseSetupBindingProfile,
  blueprint: BaseSetupCompositionBlueprint | undefined,
  bindingProfileMap: Map<string, BaseSetupBindingProfile>,
): SceneAppType {
  const blueprintFamilies = resolveDistinctBindingFamilies(
    blueprint,
    bindingProfileMap,
  );
  if (blueprintFamilies.length > 1) {
    return "hybrid";
  }

  switch (normalizeCompatSceneAppBindingFamily(binding.bindingFamily)) {
    case "browser_assist":
      return "browser_grounded";
    case "automation_job":
      return "local_durable";
    case "native_skill":
    case "agent_turn":
    default:
      return "local_instant";
  }
}

function inferPatternPrimary(
  binding: BaseSetupBindingProfile,
  artifact: BaseSetupArtifactProfile,
  blueprint: BaseSetupCompositionBlueprint | undefined,
  requiredSlotCount: number,
): SceneAppPattern {
  if ((blueprint?.steps?.length ?? 0) > 1) {
    return "pipeline";
  }
  if (binding.bindingFamily === "browser_assist") {
    return "tool_wrapper";
  }
  if (requiredSlotCount > 1) {
    return "inversion";
  }
  if (
    artifact.viewerKind === "document" ||
    artifact.deliveryContract === "table_report" ||
    artifact.deliveryContract === "project_pack"
  ) {
    return "generator";
  }
  return "pipeline";
}

function inferPatternStack(params: {
  primary: SceneAppPattern;
  binding: BaseSetupBindingProfile;
  artifact: BaseSetupArtifactProfile;
  blueprint: BaseSetupCompositionBlueprint | undefined;
  requiredSlotCount: number;
}): SceneAppPattern[] {
  const patterns = new Set<SceneAppPattern>([params.primary]);

  if ((params.blueprint?.steps?.length ?? 0) > 1) {
    patterns.add("pipeline");
  }
  if (params.requiredSlotCount > 0) {
    patterns.add("inversion");
  }
  if (params.binding.bindingFamily === "browser_assist") {
    patterns.add("tool_wrapper");
  }
  if (
    params.artifact.viewerKind === "document" ||
    params.artifact.deliveryContract === "artifact_bundle" ||
    params.artifact.deliveryContract === "project_pack" ||
    params.artifact.deliveryContract === "table_report"
  ) {
    patterns.add("generator");
  }

  return Array.from(patterns);
}

function inferInfraProfile(
  binding: BaseSetupBindingProfile,
  artifact: BaseSetupArtifactProfile,
  blueprint: BaseSetupCompositionBlueprint | undefined,
): string[] {
  const bindingFamily = normalizeCompatSceneAppBindingFamily(
    binding.bindingFamily,
  );
  const infra = new Set<string>(
    (binding.capabilityRefs ?? []).map((capabilityRef) =>
      normalizeCompatSceneAppCapabilityRef(capabilityRef),
    ),
  );
  infra.add(bindingFamily);

  if (bindingFamily === "browser_assist") {
    infra.add("browser_connector");
  }
  if (
    bindingFamily === "automation_job" ||
    binding.runnerType === "scheduled"
  ) {
    infra.add("automation_schedule");
  }
  if (
    artifact.deliveryContract === "artifact_bundle" ||
    artifact.deliveryContract === "project_pack" ||
    artifact.outputDestination
  ) {
    infra.add("workspace_storage");
  }
  if (artifact.deliveryContract === "artifact_bundle") {
    infra.add("artifact_bundle");
  }
  if (artifact.deliveryContract === "project_pack") {
    infra.add("project_pack");
  }
  if (artifact.viewerKind === "document") {
    infra.add("document_viewer");
  }
  if (artifact.viewerKind === "table_report") {
    infra.add("table_report_viewer");
  }
  if (blueprint?.id) {
    infra.add("composition_blueprint");
  }

  return Array.from(infra);
}

function buildLaunchRequirements(
  binding: BaseSetupBindingProfile,
  projection: BaseSetupCatalogProjection,
  requiredSlotCount: number,
): SceneAppLaunchRequirement[] {
  const requirements: SceneAppLaunchRequirement[] = [];

  if (requiredSlotCount > 0) {
    requirements.push({
      kind: "user_input",
      message: "需要先补齐场景输入或必填槽位。",
    });
  }
  if (projection.readinessRequirements?.requiresProject) {
    requirements.push({
      kind: "project",
      message: "该场景需要先绑定项目或工作区资产目录。",
    });
  }
  if (
    projection.readinessRequirements?.requiresBrowser ||
    normalizeCompatSceneAppBindingFamily(binding.bindingFamily) ===
      "browser_assist"
  ) {
    requirements.push({
      kind: "browser_session",
      message: "该场景需要可用的浏览器上下文或已附着会话。",
    });
  }
  if (
    normalizeCompatSceneAppBindingFamily(binding.bindingFamily) ===
      "automation_job" ||
    binding.runnerType === "scheduled"
  ) {
    requirements.push({
      kind: "automation",
      message: "该场景需要自动化调度能力。",
    });
  }

  return requirements;
}

function buildEntryBindings(params: {
  bindingFamily: BaseSetupAllowedBindingFamily;
  serviceSkillProjection?: BaseSetupCatalogProjection;
  sceneProjection?: BaseSetupCatalogProjection;
}): SceneAppEntryBinding[] {
  const bindings: SceneAppEntryBinding[] = [];
  const bindingFamily = normalizeCompatSceneAppBindingFamily(
    params.bindingFamily,
  );

  if (params.serviceSkillProjection) {
    bindings.push({
      kind: "service_skill",
      bindingFamily,
      serviceSkillId: params.serviceSkillProjection.id,
      skillKey: params.serviceSkillProjection.skillKey,
      aliases: params.serviceSkillProjection.aliases,
    });
  }

  const sceneBinding =
    params.sceneProjection?.sceneBinding ??
    params.serviceSkillProjection?.sceneBinding;
  if (sceneBinding) {
    bindings.push({
      kind: "scene",
      bindingFamily,
      sceneKey: sceneBinding.sceneKey,
      commandPrefix: sceneBinding.commandPrefix,
      aliases: sceneBinding.aliases,
      skillKey:
        params.sceneProjection?.skillKey ??
        params.serviceSkillProjection?.skillKey,
    });
  }

  return bindings;
}

function buildDeliveryProfile(params: {
  artifactProfile: BaseSetupArtifactProfile;
  compositionBlueprint?: BaseSetupCompositionBlueprint;
}): SceneAppDeliveryProfile {
  const requiredParts = dedupeStrings([
    ...(params.compositionBlueprint?.deliveryContract?.requiredParts ?? []),
    ...params.artifactProfile.requiredParts,
  ]);

  return {
    artifactProfileRef: params.artifactProfile.id,
    viewerKind: params.artifactProfile.viewerKind,
    requiredParts,
    primaryPart: requiredParts[0],
  };
}

function buildCompositionProfile(
  blueprint: BaseSetupCompositionBlueprint | undefined,
  bindingProfileMap: Map<string, BaseSetupBindingProfile>,
): SceneAppCompositionProfile | undefined {
  if (!blueprint?.id && !blueprint?.steps?.length) {
    return undefined;
  }

  const steps: SceneAppCompositionStepDescriptor[] = (
    blueprint?.steps ?? []
  ).map((step, index) => ({
    id: step.id,
    order: index + 1,
    bindingProfileRef: step.bindingProfileRef,
    bindingFamily: step.bindingProfileRef
      ? (() => {
          const bindingFamily = bindingProfileMap.get(
            step.bindingProfileRef,
          )?.bindingFamily;
          return bindingFamily
            ? normalizeCompatSceneAppBindingFamily(bindingFamily)
            : undefined;
        })()
      : undefined,
  }));

  return {
    blueprintRef: blueprint?.id,
    stepCount: steps.length,
    steps,
  };
}

function buildScorecardProfile(
  profileRef: string | undefined,
  scorecardProfile: BaseSetupScorecardProfile | undefined,
): SceneAppScorecardProfile | undefined {
  if (!profileRef && !scorecardProfile) {
    return undefined;
  }

  return {
    profileRef: profileRef ?? scorecardProfile?.id,
    metricKeys: scorecardProfile?.metrics ?? [],
    failureSignals: scorecardProfile?.failureSignals ?? [],
  };
}

function compileSceneAppDescriptor(params: {
  pkg: BaseSetupPackage;
  serviceSkillProjection?: BaseSetupCatalogProjection;
  sceneProjection?: BaseSetupCatalogProjection;
  bindingProfileMap: Map<string, BaseSetupBindingProfile>;
  artifactProfileMap: Map<string, BaseSetupArtifactProfile>;
  compositionBlueprintMap: Map<string, BaseSetupCompositionBlueprint>;
  scorecardProfileMap: Map<string, BaseSetupScorecardProfile>;
}): SceneAppDescriptor | null {
  const primaryProjection =
    params.sceneProjection ?? params.serviceSkillProjection ?? null;
  if (!primaryProjection) {
    return null;
  }

  const bindingProfile = params.bindingProfileMap.get(
    primaryProjection.bindingProfileRef,
  );
  const artifactProfile = params.artifactProfileMap.get(
    primaryProjection.artifactProfileRef,
  );
  if (!bindingProfile || !artifactProfile) {
    return null;
  }

  const slotProfile = params.pkg.slotProfiles.find(
    (profile) => profile.id === primaryProjection.slotProfileRef,
  );
  const requiredSlotCount =
    slotProfile?.slots.filter((slot) => slot.required).length ?? 0;
  const compositionBlueprint = primaryProjection.compositionBlueprintRef
    ? params.compositionBlueprintMap.get(
        primaryProjection.compositionBlueprintRef,
      )
    : undefined;
  const scorecardProfile = params.scorecardProfileMap.get(
    primaryProjection.scorecardProfileRef,
  );

  const sceneappType = inferSceneAppType(
    bindingProfile,
    compositionBlueprint,
    params.bindingProfileMap,
  );
  const patternPrimary = inferPatternPrimary(
    bindingProfile,
    artifactProfile,
    compositionBlueprint,
    requiredSlotCount,
  );
  const deliveryProfile = buildDeliveryProfile({
    artifactProfile,
    compositionBlueprint,
  });
  const compositionProfile = buildCompositionProfile(
    compositionBlueprint,
    params.bindingProfileMap,
  );
  const compiledScorecardProfile = buildScorecardProfile(
    primaryProjection.scorecardProfileRef,
    scorecardProfile,
  );

  return {
    id:
      normalizeString(primaryProjection.skillKey) ??
      normalizeString(primaryProjection.sceneBinding?.sceneKey) ??
      primaryProjection.entryKey,
    title: params.sceneProjection?.title ?? primaryProjection.title,
    summary: params.sceneProjection?.summary ?? primaryProjection.summary,
    category: primaryProjection.category,
    sceneappType,
    patternPrimary,
    patternStack: inferPatternStack({
      primary: patternPrimary,
      binding: bindingProfile,
      artifact: artifactProfile,
      blueprint: compositionBlueprint,
      requiredSlotCount,
    }),
    capabilityRefs: (bindingProfile.capabilityRefs ?? []).map((capabilityRef) =>
      normalizeCompatSceneAppCapabilityRef(capabilityRef),
    ),
    infraProfile: inferInfraProfile(
      bindingProfile,
      artifactProfile,
      compositionBlueprint,
    ),
    deliveryContract:
      artifactProfile.deliveryContract as SceneAppDeliveryContract,
    artifactKind: artifactProfile.defaultArtifactKind,
    outputHint: primaryProjection.outputHint,
    entryBindings: buildEntryBindings({
      bindingFamily: bindingProfile.bindingFamily,
      serviceSkillProjection: params.serviceSkillProjection,
      sceneProjection: params.sceneProjection,
    }),
    launchRequirements: buildLaunchRequirements(
      bindingProfile,
      primaryProjection,
      requiredSlotCount,
    ),
    linkedServiceSkillId: params.serviceSkillProjection?.id,
    linkedSceneKey:
      params.sceneProjection?.sceneBinding?.sceneKey ??
      params.serviceSkillProjection?.sceneBinding?.sceneKey,
    deliveryProfile,
    compositionProfile,
    scorecardProfile: compiledScorecardProfile,
    aliases: dedupeStrings([
      ...(primaryProjection.aliases ?? []),
      ...(params.serviceSkillProjection?.aliases ?? []),
      ...(params.sceneProjection?.aliases ?? []),
      ...(params.serviceSkillProjection?.sceneBinding?.aliases ?? []),
      ...(params.sceneProjection?.sceneBinding?.aliases ?? []),
    ]),
    sourcePackageId: params.pkg.id,
    sourcePackageVersion: params.pkg.version,
  };
}

export function compileSceneAppCatalogFromPackage(
  pkg: BaseSetupPackage,
  options: { generatedAt?: string } = {},
): SceneAppCatalog {
  const bindingProfileMap = buildBindingProfileMap(pkg);
  const artifactProfileMap = buildArtifactProfileMap(pkg);
  const compositionBlueprintMap = buildCompositionBlueprintMap(pkg);
  const scorecardProfileMap = buildScorecardProfileMap(pkg);
  const items = Array.from(buildProjectionGroups(pkg).values())
    .map((group) =>
      compileSceneAppDescriptor({
        pkg,
        serviceSkillProjection: group.serviceSkillProjection,
        sceneProjection: group.sceneProjection,
        bindingProfileMap,
        artifactProfileMap,
        compositionBlueprintMap,
        scorecardProfileMap,
      }),
    )
    .filter((item): item is SceneAppDescriptor => Boolean(item))
    .sort((left, right) => left.title.localeCompare(right.title, "zh-CN"));

  return {
    version: pkg.version,
    generatedAt: options.generatedAt ?? new Date().toISOString(),
    items,
  };
}

export function compileSceneAppCatalogFromSnapshot(
  snapshot: StoredBaseSetupPackageSnapshot,
): SceneAppCatalog {
  return compileSceneAppCatalogFromPackage(snapshot.package, {
    generatedAt: snapshot.syncedAt,
  });
}

export function readStoredSceneAppCatalog(): SceneAppCatalog | null {
  const snapshot = readStoredBaseSetupPackageSnapshot();
  if (!snapshot) {
    return null;
  }

  return compileSceneAppCatalogFromSnapshot(snapshot);
}
